etcdserver/cindex/cindex.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cindex
+
+import (
+	"encoding/binary"
+	"sync"
+	"sync/atomic"
+
+	"go.etcd.io/etcd/mvcc/backend"
+)
+
+var (
+	metaBucketName = []byte("meta")
+
+	consistentIndexKeyName = []byte("consistent_index")
+)
+
+// ConsistentIndexer is an interface that wraps the Get/Save method.
+// Consistent index is the offset of an entry in a consistent replicated log.
+type ConsistentIndexer interface {
+
+	// ConsistentIndex returns the consistent index of current executing entry.
+	ConsistentIndex() uint64
+
+	// SetConsistentIndex set the consistent index of current executing entry.
+	SetConsistentIndex(v uint64)
+
+	// UnsafeSave must be called holding the lock on the tx.
+	// It saves consistentIndex to the underlying stable storage.
+	UnsafeSave(tx backend.BatchTx)
+
+	// SetBatchTx set the available backend.BatchTx for ConsistentIndexer.
+	SetBatchTx(tx backend.BatchTx)
+}
+
+// consistentIndex represents the offset of an entry in a consistent replica log.
+// It implements the ConsistentIndexer interface.
+// It is always set to the offset of current entry before executing the entry,
+// so ConsistentWatchableKV could get the consistent index from it.
+
+type consistentIndex struct {
+	tx backend.BatchTx
```

Comment:
One more question. Do we need to move tx below consistentIndex to keep consistentIndex 64-bit aligned?

Reply:
sizeof(interface) is 8 bytes in 32 bit,16 bytes in 64 bit. so we do do need to move tx below consistentIndex.  
https://play.golang.org/p/9Bj6zdtyehz